Added the cmake-config command

Tyler-g-hudson opened the pull request:
+ also added a defaults.py file that holds default names and paths.
~ Moved the default tag prefix to the defaults file.

Tyler-g-hudson committed:
Added the cmake-config command

+ also added a defaults.py file that holds default names and paths.
~ Moved the default tag prefix to the defaults file.

Tyler-g-hudson committed:
Removed the moved universal tag function from _utils.py

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+    with_cuda: bool = True,
+) -> str:
+    """
+    Creates a dockerfile for configuring CMAKE.
```
```suggestion
    Creates a Dockerfile for configuring CMake build.
```

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+    base : str
+        The base image tag.
+    build_type : str
+        The CMAKE build type.
```
```suggestion
        The CMake build type. See 
        `here <https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html>`_
        for possible values.
```

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+        The base image tag.
+    build_type : str
+        The CMAKE build type.
+    with_cuda : bool
```
```suggestion
    with_cuda : bool, optional
```

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+    dockerfile += micromamba_docker_lines() + "\n\n"
+    dockerfile += dedent(
+        f"""
+            ENV INSTALL_PREFIX {str(install_prefix())}
+            ENV BUILD_PREFIX {str(build_prefix())}
```
Do these environment variables get used later on at all?

It seems non-ideal that they get defined here as a side-effect of the CMake configuration step, and aren't documented at all.

gmgunter replied:
If these variables get used later on, the best solution might just be to document them here -- i.e. update the docstring to clarify that these variables are defined in this step.

Another alternative might be to pass the `install_prefix` and `build_prefix` as inputs to this function and default them to the values in `defaults.py`. Then you presumably wouldn't need the environment variables. Instead, the user would just need to make sure that they pass the same `install_prefix` and `build_prefix` paths to later steps that they passed to this step.

I think that latter alternative might be my preferred approach, actually. I wouldn't necessarily expose `install_prefix` and `build_prefix` all the way up to the command line interface, but for this low-level function I think it might be nice for them to be passed as (optional) input parameters. And any other low-level functions that rely on the same `install_prefix` and/or `build_prefix` would also take them as input parameters. I think that would help to avoid a "hidden dependency" between these functions.

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+        f"""
+            ENV INSTALL_PREFIX {str(install_prefix())}
+            ENV BUILD_PREFIX {str(build_prefix())}
+            ENV PYTHONPATH $INSTALL_PREFIX/packages:$PYTHONPATH
```
This is prepending the Python package installation path to the `PYTHONPATH` environment variable. I think it makes more sense to do this later on when we actually install the software. Right now we're just doing the CMake configuration step.

```suggestion
```

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+                -DCMAKE_PREFIX_PATH=$MAMBA_ROOT_PREFIX \\
+                -DWITH_CUDA=YES \\
+                {cmake_extra_args} \\
+                .
```
We're passing `.` as a positional argument to `cmake` here, which means that we're assuming that the project source directory (that contains the top-level `CMakeLists.txt` file) is the current working directory. This isn't currently documented anywhere -- it's just an implicit assumption.

I think the code would be more explicit and clear if we remove this assumption and instead make the source directory path a named variable that we pass to CMake using the `-S` flag, 

```diff
            RUN cmake \\
+                -S {src_prefix} \\
                 -B $BUILD_PREFIX \\
                 -G Ninja \\
                 -DISCE3_FETCH_DEPS=NO \\
                 -DCMAKE_BUILD_TYPE={build_type} \\
                 -DCMAKE_INSTALL_PREFIX=$INSTALL_PREFIX \\
                 -DCMAKE_PREFIX_PATH=$MAMBA_ROOT_PREFIX \\
                 -DWITH_CUDA=YES \\
-                {cmake_extra_args} \\
-                .
+                {cmake_extra_args}
```

Where `src_prefix` is either hard-coded in the function body like `INSTALL_PREFIX` and `BUILD_PREFIX` are currently, or else specified by the user as an input argument to this function.

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+                -DCMAKE_BUILD_TYPE={build_type} \\
+                -DCMAKE_INSTALL_PREFIX=$INSTALL_PREFIX \\
+                -DCMAKE_PREFIX_PATH=$MAMBA_ROOT_PREFIX \\
+                -DWITH_CUDA=YES \\
```
The `-DWITH_CUDA` flag is already included in `cmake_extra_args` so we don't want to specify it here as well.

```suggestion
```

gmgunter replied:
Let's add a space between the `-D` part and the `<key>=<value>` part of these command line arguments for aesthetics/readability and consistency with the `-B` and `-G` arguments above.

```suggestion
                -D ISCE3_FETCH_DEPS=NO \\
                -D CMAKE_BUILD_TYPE={build_type} \\
                -D CMAKE_INSTALL_PREFIX=$INSTALL_PREFIX \\
                -D CMAKE_PREFIX_PATH=$MAMBA_ROOT_PREFIX \\
                -D WITH_CUDA=YES \\
```

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,61 @@
+    additional_args = []
+    if with_cuda:
+        additional_args += ["-DWITH_CUDA=YES"]
+    cmake_extra_args = " ".join(additional_args)
```
FYI if you don't explicitly pass `-DWITH_CUDA=YES` to the isce3 cmake build, it will check to see if your system has CUDA installed and will either enable or disable the isce3-cuda library components depending on whether CUDA was found or not. If you definitely don't want CUDA, you can explicitly say `-DWITH_CUDA=NO` in order to disable this behavior.

I think the best thing to do here is just to keep it simple and just support the two cases where CUDA is explicitly enabled/disabled:

```suggestion
    additional_args = []
    if with_cuda:
        additional_args += ["-D WITH_CUDA=YES"]
    else:
        additional_args += ["-D WITH_CUDA=NO"]
    cmake_extra_args = " ".join(additional_args)
```

But, alternatively, you could sort of replicate the behavior of the cmake scripts here by allowing the user to choose `with_cuda=None` in addition to `True` or `False` and then implement something like this:

```suggestion
    additional_args = []
    if with_cuda is None:
        pass
    elif with_cuda:
        additional_args += ["-D WITH_CUDA=YES"]
    else:
        additional_args += ["-D WITH_CUDA=NO"]
    cmake_extra_args = " ".join(additional_args)
```

But that might be more complexity than you want to add here. In particular, it seems like the command line argument would be kinda complicated because you couldn't just use `action="store_true"` in that case. So I would probably just stick with the simple True/False option.

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -39,6 +39,24 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        "--build-type",
+        type=str,
+        default="Release",
+        metavar="DCMAKE_BUILD_TYPE",
```
Let's just call this "CMAKE_BUILD_TYPE". The "-D" part in the beginning is just a cmake command line flag -- it's not actually part of the variable name.

```suggestion
        metavar="CMAKE_BUILD_TYPE",
```

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -39,6 +39,24 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        default="Release",
+        metavar="DCMAKE_BUILD_TYPE",
+        choices=build_type_choices,
+        help="The --DCMAKE_BUILD_TYPE argument for CMAKE. Valid options are: "
```
```suggestion
        help="The CMAKE_BUILD_TYPE argument for CMake. Valid options are: "
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -149,6 +146,45 @@ def copy_dir(
     )
 
 
+def configure_cmake(
```
I don't see any tests exercising this function yet.

Let's add a test where we run this step and then check that the build directory exists and contains a file named `CMakeCache.txt` in the resulting image. For bonus points, you could parse the `CMakeCache.txt` file and check that the values of `CMAKE_BUILD_TYPE` and `WITH_CUDA` match the expected values.

Tyler-g-hudson replied:
Done, with bonus points.

gmgunter reviewed (commented):
Thanks! Looks good overall. Few suggestions below

Tyler-g-hudson committed:
src prefix, CMAKE -> CMake,  docs changes

+ Added src_prefix function to defaults.
~ Changed references to CMAKE to instead say CMake
~ Some documentation improvement

Tyler-g-hudson committed:
Small doc change

Tyler-g-hudson committed:
Added unit tests for CMake config.

+ Added a cmake mark.
+ Added fixtures, functions, and tests for testing CMake config.
+ Added a CMake and Ninja lock file.

Tyler-g-hudson committed:
BUGFIX: Unintended removal of init images by testing

Normal init images were removed by testing.
+ Added a function for making init image Dockerfiles.
~ Moved the above funtionality out of setup_init().
~ Changed the init_image fixture to add a unique directory to the init dockerfile so it will not delete images unintentionally.

Tyler-g-hudson commented on docker_cli/_docker_init.py:
```diff
@@ -0,0 +1,37 @@
+from textwrap import dedent
```
@gmgunter changes made in this file

Tyler-g-hudson commented on test/fixtures.py:
```diff
@@ -92,7 +92,20 @@ def init_image(base_tag: str, init_tag: str) -> Iterator[Image]:
         The initialization tag generator.
     """
-    img, _, _ = setup_init(base=base_tag, tag=init_tag, no_cache=False)
+    # Get some initial install lines to ensure that the appropriate software is
+    # installed on the init image.
+    with temp_image(base_tag) as temp_img:
+        _, _, initial_lines = image_command_check(temp_img, True)
+
+    # Get the init image dockerfile.
+    dockerfile = init_dockerfile(base=base_tag, custom_lines=initial_lines)
+
+    # Creates a unique directory in the image. This causes the image to be unique,
+    # preventing unintentional deletion of init images that are not for testing.
+    dockerfile += f"\n\nRUN mkdir {init_tag}"
+
+    # Build and yield the image.
+    img = Image.build(tag=init_tag, dockerfile_string=dockerfile)
```
@gmgunter hotfix changes made here

Tyler-g-hudson commented on docker_cli/setup_commands.py:
```diff
@@ -42,26 +39,9 @@ def setup_init(
-    )
+        package_mgr, url_reader, initial_lines = image_command_check(temp_img, True)
+
+    dockerfile = init_dockerfile(base=base, custom_lines=initial_lines)
```
@gmgunter hotfix changes made here

Tyler-g-hudson reviewed (commented):
This review is intended to assist in locating the largest of the changes I made in my hotfix commit.

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,59 @@
+            ENV BUILD_PREFIX {str(build_prefix())}
+
+            RUN cmake \\
+                -S {src_prefix()} \\
```
```suggestion
                -S . \\
```

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -0,0 +1,59 @@
+
+def cmake_config_dockerfile(base: str, build_type: str, with_cuda: bool = True) -> str:
+    """
+    Creates a Dockerfile for configuring CMake Build.
```
Let's add a note that mentions that we assume that the current working directory is the source directory (that contains the top-level `CMakeLists.txt` file).

gmgunter commented on docker_cli/defaults.py:
```diff
@@ -0,0 +1,21 @@
+    """Returns the build system's default build prefix path."""
+    return Path("/tmp/build")
+
+
+def src_prefix() -> str:
+    """Returns the filesystem prefix to the source code."""
+    return "."
```
```suggestion
```

Tyler-g-hudson committed:
Removed src_prefix()

gmgunter reviewed (approved):
Looks great, thanks!
